ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -805,19 +861,39 @@ void handleBody(InputStream payload)
               MetadataContentSerDe.deserializeMetadataContentRecord(serializedMetadataContent, blobIdFactory);
           chunkSize = compositeBlobInfo.getChunkSize();
           totalSize = compositeBlobInfo.getTotalSize();
-          List<StoreKey> keys = compositeBlobInfo.getKeys();
-          chunkIdIterator = keys.listIterator();
-          numChunksTotal = keys.size();
-          dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
-          for (int i = 0; i < dataChunks.length; i++) {
-            dataChunks[i] = new GetChunk(chunkIdIterator.nextIndex(), (BlobId) chunkIdIterator.next());
+          try {
+            List<StoreKey> keys = compositeBlobInfo.getKeys();
+            if (hasRange) {
+              validatedByteRange = new ValidatedByteRange(options.getRange(), totalSize);
+              // Get only the chunks within the range.
+              firstChunkIndexInRange = (int) (validatedByteRange.getStartOffset() / chunkSize);
+              int lastChunkIndexInByteRange = (int) (validatedByteRange.getEndOffset() / chunkSize);
+              keys = keys.subList(firstChunkIndexInRange, lastChunkIndexInByteRange + 1);
+            }
+            chunkIdIterator = keys.listIterator();
+            numChunksTotal = keys.size();
+            dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
+            for (int i = 0; i < dataChunks.length; i++) {
+              dataChunks[i] = new GetChunk(chunkIdIterator.nextIndex(), (BlobId) chunkIdIterator.next());
+            }
+          } catch (InvalidByteRangeException e) {
+            onInvalidRange(e.getMessage());
           }
         } else {
-          chunkIdIterator = null;
-          numChunksTotal = 1;
-          dataChunks = null;
-          chunkIndexToBuffer.put(0, blobData.getStream().getByteBuffer());
-          numChunksRetrieved = 1;
+          totalSize = blobData.getSize();
+          try {
+            if (hasRange) {
+              validatedByteRange = new ValidatedByteRange(options.getRange(), totalSize);
+            }
+            chunkIdIterator = null;
+            dataChunks = null;
+            chunkIndex = 0;
+            numChunksTotal = 1;
+            chunkIndexToBuffer.put(0, filterChunkToRange(blobData));
+            numChunksRetrieved = 1;
+          } catch (InvalidByteRangeException e) {
+            onInvalidRange(e.getMessage());
```

Comment:
not sure why we need to throw an `Exception`, catch it and then handle the invalid case. What do you think about calling this function as soon as you detect a bad range (which you can do if you get rid of `ValidatedByteRange`)? Instead of creating an exception, throwing it and then handling it.